pvlib/modelchain.py

```diff
@@ -1238,29 +1505,29 @@ def _prepare_temperature(self, data=None):
         -------
         self
 
-        Assigns attribute ``cell_temperature``.
+        Assigns attribute ``results.cell_temperature``.
 
         """
-        if 'cell_temperature' in data:
-            self.cell_temperature = data['cell_temperature']
-            return self
```

Comment:
docstring says

> If 'data' contains 'cell_temperature', these values are assigned to attribute ``cell_temperature``.

and here the methods returns before doing anything else. 

the new method does assign the cell temperature, but it does not return before doing anything else. (at least for num_arrays > 1 - didn't test the old behavior) As a result, it's no longer possible to run `run_model_from_effective_irradiance` with only keys `effective_irradiance` and `cell_temperature`.

Reply:
I'll add a test for this case. I'll need to re-think the way we handle missing `cell_temperature` in some, but not all of the input data frames. As it stands I just compute it for everything and don't use the ones we don't need. Good opportunity to get rid of that inefficiency.

Reply:
I'm now remembering why I did it this way to begin with. We currently compute cell temperature via the `PVSystem` method for the given cell temperature model. This doesn't allow the option for only computing it on some of the arrays. If we want to do that we should move the meat of the `PVSystem.xxxx_celltemp` functions to `Array.xxxx_celltemp`. This would allow calculation of cell temperature for each array separately.

I think we should do this regardless of how the above capability is implemented, since cell temperature is inherently a per-array quantity. (We may also want to move `PVSystem.sapm_effective_irradiance()` and `PVSystem.xxxx_spectral_loss()` functions into `Array`, but that isn't strictly necessary at this point.)

Reply:
>  This would allow calculation of cell temperature for each array separately.

Are you saying that in the current code, a single cell temperature is computed for the system? I thought `PVSystem.xxxx_celltemp` would return a tuple, one element per Array.

Reply:
No, the cell temperature varies between arrays, but the API only lets you compute it for all arrays at the same time. What I need is something like `system.arrays[0].xxxx_celltemp()`.

Reply:
What is the downside of moving `xxxx_celltemp` to Array? We could retain `PVSystem.xxxx_celltemp` as a wrapper through a deprecation period. In fact this may give us a chance to rename those cell temperature methods @wholmgren

Reply:
Two costs are immediately apparent: it is one more layer of indirection and it increases the size of the `Array` API. I think both are worth it since it will also substantially simplify the `PVSystem` methods which essentially become responsible for nothing more than input validation and iteration over the `Arrays`.

It may cause increased complexity in `ModelChain` if we need to keep one system-wide function assigned to `ModelChain.temperature_model` and one internal per-array function. However, we might be able mitigate that by refactoring to a single `cell_temperature` function with a keyword arg.

Reply:
+1 for a `Array.get_cell_temperature`, but I'm not convinced it's needed in this PR. 

I'm fine with requiring all input dataframes to use the same temperature data. I just don't want to lose the ability to run the model chain with only `cell_temperature`. 

I think you can do something like

```python
        if all(x is not None for x in given_cell_temperature):
            return given_cell_temperature

        # wasn't supplied with cell or module temperature 
        # for at least one array, so calculating it.
        self.temperature_model()
        # replace calculated cell temperature with temperature given in `data`
        # where available.
        self.results.cell_temperature = tuple(
            itertools.starmap(
                lambda given, modeled: modeled if given is None else given,
                zip(given_cell_temperature, self.results.cell_temperature)
            )
        )
```

Reply:
I updated it to use the early return you suggest. It works, but I'm not really liking the idea of requiring input data that won't be used ('poa_global' for Arrays that already have 'cell_temperature'). If we don't want to go the route of adding `Array.get_cell_temperature` in this PR we can maybe get around it by filling in "dummy" data for arrays that need it. Just have to be careful that the output for those arrays is not assigned in `ModelChain`.

Reply:
-1 to my previous suggestion. I tried to flesh it out and it gets real messy real quick.

Reply:
> It works, but I'm not really liking the idea of requiring input data that won't be used ('poa_global' for Arrays that already have 'cell_temperature').

I'm confused. I'm trying to avoid requiring `poa_global` for Arrays that already have `cell_temperature`. Are you saying this introduces a dependence?

Reply:
Rather than a very difficult to understand `KeyError` let's check that `'poa_global'` exists for all `Arrays` and raise a useful & informative error message if it does not.

Reply:
But going back to my original concern... `run_model_from_effective_irradiance` should work with only `'effective_irradiance'` and `'cell_temperature'` - no need for `'poa_global'`.

Reply:
> But going back to my original concern... `run_model_from_effective_irradiance` should work with only `'effective_irradiance'` and `'cell_temperature'` - no need for `'poa_global'`.

I don't quite agree - `run_model_from_effective_irradiance` should also work using a cell temperature model. Where `'effective_irradiance'` is provided I think it is OK to use those values in place of `'poa_global'` for calculating cell temperature. OK with me if that capability is in a later PR.

Reply:
> run_model_from_effective_irradiance should also work using a cell temperature model. 

Should *also* work, yes. My concern is that it's also no longer possible to work with only `cell_temperature` and `effective_irradiance`. Looking at the code now, I think the issue is resolved, it just wasn't clear in the discussion.

> Where 'effective_irradiance' is provided I think it is OK to use those values in place of 'poa_global' for calculating cell temperature. OK with me if that capability is in a later PR.

Agree, agree. I thought you considered and rejected this in when you implemented the method, but perhaps I'm misremembering.